File: paradigms/factorial_fibonacci.ts
/**
 * Factorial & Fibonacci Analysis
 * 
 * Factorial:
 * - Recursive: O(n) time, O(n) space (stack)
 * - Iterative: O(n) time, O(1) space
 * 
 * Fibonacci:
 * - Recursive: O(2^n) time (without memo), O(n) space
 * - Iterative: O(n) time, O(1) space
 */

// --- Factorial ---

function factorialRecursive(n: number): number {
    if (n <= 1) return 1; // Base case
    return n * factorialRecursive(n - 1); // Recursive step
}

function factorialIterative(n: number): number {
    let result = 1;
    for (let i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

// --- Fibonacci ---

function fibonacciRecursive(n: number): number {
    if (n <= 1) return n;
    return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

function fibonacciIterative(n: number): number {
    if (n <= 1) return n;
    let prev = 0, curr = 1;
    for (let i = 2; i <= n; i++) {
        let next = prev + curr;
        prev = curr;
        curr = next;
    }
    return curr;
}

// --- Demo ---

const nFactorial = 5;
const nFibonacci = 10;

console.log(`--- Factorial of ${nFactorial} ---`);
console.log(`Recursive: ${factorialRecursive(nFactorial)}`);
console.log(`Iterative: ${factorialIterative(nFactorial)}`);

console.log(`\n--- Fibonacci of ${nFibonacci} ---`);
console.log(`Recursive: ${fibonacciRecursive(nFibonacci)}`);
console.log(`Iterative: ${fibonacciIterative(nFibonacci)}`);

File: paradigms/factorial_fibonacci.rs
/**
 * Recursion vs Iteration in Rust
 */

// --- Factorial ---

fn factorial_recursive(n: u64) -> u64 {
    if n <= 1 {
        return 1;
    }
    n * factorial_recursive(n - 1)
}

fn factorial_iterative(n: u64) -> u64 {
    let mut result = 1;
    for i in 2..=n {
        result *= i;
    }
    result
}

// --- Fibonacci ---

fn fibonacci_recursive(n: u32) -> u32 {
    if n <= 1 {
        return n;
    }
    fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)
}

fn fibonacci_iterative(n: u32) -> u32 {
    if n <= 1 {
        return n;
    }
    let mut prev = 0;
    let mut curr = 1;
    for _ in 2..=n {
        let next = prev + curr;
        prev = curr;
        curr = next;
    }
    curr
}

fn main() {
    let n_factorial = 5;
    let n_fibonacci = 10;

    println!("--- Factorial of {} ---", n_factorial);
    println!("Recursive: {}", factorial_recursive(n_factorial));
    println!("Iterative: {}", factorial_iterative(n_factorial));

    println!("\n--- Fibonacci of {} ---", n_fibonacci);
    println!("Recursive: {}", fibonacci_recursive(n_fibonacci));
    println!("Iterative: {}", fibonacci_iterative(n_fibonacci));
}
